### PR TITLE
Fix Some Leaks

### LINKED ISSUE #2098: custom function leak
## versions of stuff

- NPM version (`npm -v`):  `5.1.0`
- Node version (`node -v`): `v8.1.3`
- Node Process (`node -p process.versions`):
```
{ http_parser: '2.7.0',
  node: '8.1.3',
  v8: '5.8.283.41',
  uv: '1.12.0',
  zlib: '1.2.11',
  ares: '1.10.1-DEV',
  modules: '57',
  openssl: '1.0.2l',
  icu: '59.1',
  unicode: '9.0',
  cldr: '31.0.1',
  tz: '2017b' }
```
- Node Platform (`node -p process.platform`): `darwin`
- Node architecture (`node -p process.arch`): `x64`
- node-sass version (`node -p "require('node-sass').info"`):
```
node-sass	4.5.3	(Wrapper)	[JavaScript]
libsass  	3.5.0.beta.2	(Sass Compiler)	[C/C++
```

## Issue

Using custom functions results in unbounded growth, for example:

```js
const sass = require('.');
function renderSync() {
  sass.renderSync({
    data: '#{headings(2,5)} { color: #08c; }',
    functions: {
      'headings($from: 0, $to: 6)': function(from, to) {
        let f = from.getValue();
        let t = to.getValue();
        let list = new sass.types.List(t - f + 1);

        for (let i = f; i <= t; i++) {
          list.setValue(i - f, new sass.types.String('h' + i));
        }

        return list;
      }
    }
  });
}

while (true) {
  renderSync();
}
```

When inspecting the process (limiting to only one `renderSync` it appears \w leak is of arguments and returns values.

```js
renderSync();
```
using:

```
export MallocStackLogging=1
node <above-script>
leaks <pid-of-the-above-node-process>
```

```sh
Process:         node [91877]
Path:            /Users/spenner/src/stefanpenner/dotfiles/.config/node/versions/node-v8.1.3-darwin-x64/bin/node
Load Address:    0x100000000
Identifier:      node
Version:         ???
Code Type:       X86-64
Parent Process:  fish [86459]

Date/Time:       2017-09-14 10:58:20.742 -0700
Launch Time:     2017-09-14 10:57:47.137 -0700
OS Version:      Mac OS X 10.12.6 (16G29)
Report Version:  7
Analysis Tool:   /Applications/Xcode.app/Contents/Developer/usr/bin/leaks
Analysis Tool Version:  Xcode 8.3.3 (8E3004b)
----

leaks Report Version:  2.0
Process 91877: 2465 nodes malloced for 689 KB
Process 91877: 5 leaks for 112 total leaked bytes.
Leak: 0x102505c20  size=32  zone: DefaultMallocZone_0x10232b000  length: 26  "headings($from: 0, $to: 6)"
Leak: 0x102505c40  size=32  zone: DefaultMallocZone_0x10232b000  length: 26  "headings($from: 0, $to: 6)"
Leak: 0x10270e0c0  size=16  zone: DefaultMallocZone_0x10232b000  length: 2  "h2"
Leak: 0x10270e450  size=16  zone: DefaultMallocZone_0x10232b000  length: 2  "h4"
Leak: 0x10270e500  size=16  zone: DefaultMallocZone_0x10232b000  length: 2  "h5"
```

###  Relevant Malloc traces 

* renderFile + ExtractOptions
  [ExtractOptions](https://github.com/sass/node-sass/blob/aec7049bef2e5262083784b3fc5b6f7d97fb005c/src/binding.cpp#L35) allocates strings via `create_string` but does not free them

```
Leak: 0x101f065d0  size=48  zone: DefaultMallocZone_0x101e27000  length: 34  "eyeglass-fs-exists($absolute-path)"
        Call stack: [thread 0x7ffff04783c0]:
        0x4c93af092a7
        v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*)
        v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::(anonymous namespace)::BuiltinArguments<(v8::internal::BuiltinExtraArguments)1>)
        v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&))
        Nan::imp::FunctionCallbackWrapper(v8::FunctionCallbackInfo<v8::Value> const&)
        render_file(Nan::FunctionCallbackInfo<v8::Value> const&)
        ExtractOptions(v8::Local<v8::Object>, void*, sass_context_wrapper*, bool, bool)
        create_string(v8::MaybeLocal<v8::Value>)
        malloc
```

* custom function related:
  [pre_process_args](https://github.com/sass/node-sass/blob/master/src/custom_function_bridge.cpp#L20) appears to allocate SassTypes (and strings within them) but it appears they are not free'd

```
Leak: 0x101f078d0  size=16  zone: DefaultMallocZone_0x101e27000  length: 4  "Sans"
        Call stack: [thread 0x7ffff04783c0]:
        0x4c93af092a7
        v8::internal::Builtin_HandleApiCallConstruct(int, v8::internal::Object**, v8::internal::Isolate*)
        v8::internal::Builtin_Impl_HandleApiCallConstruct(v8::internal::(anonymous namespace)::BuiltinArguments<(v8::internal::BuiltinExtraArguments)1>, v8::internal::Isolate*)
        v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&))
        Nan::imp::FunctionCallbackWrapper(v8::FunctionCallbackInfo<v8::Value> const&)
        SassTypes::SassValueWrapper<SassTypes::String>::New(Nan::FunctionCallbackInfo<v8::Value> const&)
        SassTypes::String::construct(std::vector<v8::Local<v8::Value>, std::allocator<v8::Local<v8::Value> > >, Sass_Value**)
        create_string(v8::MaybeLocal<v8::Value>)
        malloc
```

### PR DESCRIPTION
**Pull request by stefanpenner**
[FIXES #2098]

We noticed our ember-cli builds for linkedin.com were consuming ~10gb of memory, [after some investigation](https://github.com/sass/node-sass/issues/2098) node-sass (via eyeglass) was identified as the primary culprit. With this PR, memory usage dropped by from ~10gb -> ~1gb (saving about 9gb of ram)

**NOTE: This does not fix all the leaks, more PR's to come**

### Fixes so far:

* `create_string` must be paired with a free (this catches a few of the easy ones)
  * number constructor's unit (`sass_make_number` copies the unit)
  * string constructor's value (`sass_make_string` also copies the string)
  * Bindings `ExtractOptions` function signature (`sass_make_function` copies the `signature`)
* `SassTypes::Value` are now subclasses of `Nan::ObjectWrap`, and we utilize its machinery for memory management. This ensures our wrapper objects no longer leak [Nan::ObjectWrap docs](https://github.com/nodejs/nan/blob/master/doc/object_wrappers.md#api_nan_object_wrap)

### Tests:
* add some isolated memory-test examples, which are handy to diagnose/fix memory leaks
* back-port tests for `sass.types.*`
* in sass-eyeglass/eyeglass
* in the build system for Linkedin.com

### Notes:

I realize this PR is quite large, the only drastic change was utilizing Nan::ObjectWrap memory management machinery. The rest is a mixture of tests, and cleanup to simplify the memory issues.

I look forward to you code reviews!

### CODE REVIEW

**Comment by stefanpenner** [comment]
Looks like i got some eslint errors, will circle back shortly and fix.

**Comment by saper** [comment]
Looks good to me, only need to wade through white space changes....

**Comment by stefanpenner** [comment]
> only need to wade through white space changes....

sorry about that, I can likely split the whitespace stuff into a separate commit tonight if you need.

**Comment by kyleshay** [comment]
FYI the `w=1` query param hides whitespace changes: 
https://github.com/sass/node-sass/pull/2128/files?w=1

**Comment by xzyfer** [comment]
Thanks @stefanpenner, this all looks great to me. I'll investigate what's up with CI, it failure looks to be transient.

**Comment by stefanpenner** [comment]
@xzyfer/@saper thanks. If you got some time to chat let me know. I have some ideas that might improve this bridge code some more, but would be a larger change, and I wouldn’t want to go down that path unless it sounds semi ok to you folks.

**Comment by stefanpenner** [comment]
@xzyfer I appears the failure is on node 9, its possible that https://github.com/sass/node-sass/pull/2130 may address? Will investigate locally

**Comment by xzyfer** [comment]
I see. We test against latest which has since been updates to Node 9 before we were ready.

We're definitely open to talking more about the binding code. Most of it predates myself, and possibly all the active currently team. You can [join our Slack][1] to chat (timezones permitting).

[1]: https://libsass-slack.herokuapp.com/

**Comment by stefanpenner** [comment]
rebased \w node 9 appveyor fix 🤞

**Comment by stefanpenner** [comment]
@xzyfer if this is green, do you think we can get a release out soon? If I can help at all with that, please don't hesitate to reach out.

**Comment by xzyfer** [comment]
We'll have to rebuild our binaries once this is merged which can take a
while. We'll do a release with node 9 support first because that only
requires building a binary for one version of node. This will fast follow.

**Comment by stefanpenner** [comment]
@xzyfer can you guestimate an ETA  (1 day, after the weekend, 1 week, 1 month etc). I have to coordinate some stuff on my end of the world if this is a ways out yet. This would help me out.

**Comment by xzyfer** [comment]
Either in the next 12hrs or when I get back from holidays in a week.

**Comment by stefanpenner** [comment]
@xzyfer awesome, thanks for the heads-up!

**Comment by stefanpenner** [comment]
its green!

File: src/custom_function_bridge.cpp [hunk 1]
```diff
@@ -17,7 +17,10 @@ std::vector<v8::Local<v8::Value>> CustomFunctionBridge::pre_process_args(std::ve
   std::vector<v8::Local<v8::Value>> argv = std::vector<v8::Local<v8::Value>>();
 
   for (void* value : in) {
-    argv.push_back(SassTypes::Factory::create(static_cast<Sass_Value*>(value))->get_js_object());
+    Sass_Value* x = static_cast<Sass_Value*>(value);
+    SassTypes::Value* y = SassTypes::Factory::create(x);
+
+    argv.push_back(y->get_js_object());
```
Comment (saper) [inline src/custom_function_bridge.cpp]: is this only cosmetic change or is there some deeper reason behind it?

Comment (stefanpenner) [reply src/custom_function_bridge.cpp]: cosmetic

**Comment by saper** [comment]
Merged with 734c7c8598e6b8d4a74731184adec7d6ae735d02, thank you very much!

**Comment by nschonni** [comment]
@saper I think we should revert this (temporarily) and land #2147 as a 4.6.1. Including this means we need to rebuild every binary

**Comment by stefanpenner** [comment]
> Merged with 734c7c8, thank you very much!

@saper I will likely have some more leak fixes soonish, but these address the largest of the leaks I've run into. It took our build from 10gb ram -> 1.3gb ram.

**Comment by xzyfer** [comment]
I agree with @nschonni. I'm also away atm so we can't build the new binaries.

**Comment by saper** [comment]
Temporarily reopened

**Comment by saper** [comment]
Shall we cut v4.6.1 now with  #2147 ?

**Comment by xzyfer** [comment]
I have some questions about #2147. I'm not sure it's ready.

**Comment by saper** [comment]
Some of us are in Europe, too :) thanks

**Comment by stefanpenner** [comment]
Just sharing some positive test results:

We have been running with ^ for a few days now, Thousands of builds, across various node versions and MacOS Sierra, RHEL 6, RHEL 7. No issues and just (as predicted) a nice drop in RAM usage. We have seen as much as 22gb reduction in memory usage...

**Comment by xzyfer** [comment]
I'm aiming to land this in the next 48hrs.

I'm actually hanging out  with Chris this weekend. He says  you have ideas
on how to improve the binding layer. Feel free to join us our slack to
discuss further. There's a slack badge in the read me.

**Comment by xzyfer** [comment]
Since we need to rebuild binaries for this I'm also going to bump libsass to that latest 3.5 release to fix a couple other outstanding issues.

**Comment by saper** [comment]
Great. TIL: using `git commit ----cleanup=whitespace` to add release notes in Markdown to the [release commit](https://github.com/sass/node-sass/commit/6b7b67901bc7c1f803073089ad93b95ec5fc61d4) - this will be [picked up](https://github.com/sass/node-sass/blob/6b7b67901bc7c1f803073089ad93b95ec5fc61d4/appveyor.yml#L80) by @AppVeyor and added to the release tag on GitHub.

**Comment by xzyfer** [comment]
Cool! I'll tag a new libsass release in the next 24hrs. We'll jump straight
to that version.

**Pull request merged**